docs/parser.md

```diff
@@ -0,0 +1,247 @@
+error is an object that contains a `text` and `action`).  Instead, we support only the simplest of the forms outlined
+above, as in this updated snippet:
+
+```
+"age": {
+  "minimum": 13,
+  "errors": {
+    "minimum": "Should be at least 13 years."
+  }
+},
+```
+
+The working draft is not clear on how "required" fields should be evolved to use custom errors.  This module uses
+the schema proposed [in this GitHub issue](https://github.com/json-schema/json-schema/issues/222).  Here is an example
+of how this module expects `errors` data to be entered for required fields:
+
+```
+{
+  "properties": {
+    "shallowlyRequired": {
+      "type": "string",
+      "properties": {
+        "deeplyRequired": { "type": "string" }
+      },
+      "required": ["deeplyRequired"],
+      "errors": {
+        "required/0": "This field is required and I'm telling you about it from within a field definition."
+      }
+    }
+  },
+  "required": ["shallowlyRequired"],
+  "errors": {
+    "required/0": "This field is required and I'll tell you about it at the document level.",
+    "shallowlyRequired/required/0": "This deep field is required."
+  }
+}
+```
+
+# Requirements
+
+To use this component, you will need to instantiate it and make it aware of your schemas.  These will be dereferenced
+and cached as they are added.  You can then use the `evolveError` method to resolve the custom error definition for a
+field using the schema key and the field's JSON pointer.  See the validator implementation for an example of its use.
+
+This component uses [json-schema-ref-parser](https://github.com/BigstickCarpet/json-schema-ref-parser).  We configure
+it using `options.parserOptions` (empty by default).  Read the library's documentation for details of [what options
+are available](https://github.com/BigstickCarpet/json-schema-ref-parser/blob/master/docs/options.md).
+
+Note that this library expects to be able to resolve external references relative to the first argument passed to its
+`dereference` function, and it expects to load and cache the references itself as needed.  Since we are receiving keys
+from the `validator`, we need to some how convert these to URLs.  We do this using `fluid.stringTemplate` to combine
+`options.schemaPath` with the individual key, which means:
+
+1.  If `options.schemaPath` is a directory location, a full path will be produced.
+2.  If `options.schemaPath` is a URL, a full URL will be produced.
+3.  If `options.schemaPath` is empty, paths will be resolved relative to the working directory.
+
+# Components
+
+## `gpii.schema.parser`
+
+The base component is designed to work both on the client and server side.
+
+## `gpii.schema.parser.server`
+
+A server-side version of the component which can handle package-relative `schemaPath` entries like `%gpii-package-name/path/within/package`.
+
+# Functions
+
+## gpii.schema.parser.dereferenceSchema(that, schemaPath, schemaKey)
```

Comment:
These are better advertised as component-bound facilities (via an invoker) rather than concrete public functions, since they accept a component as the first argument. Advertising them as component-bound gives the possibility for variant implementations, conforming to an abstract contract.

However, you do need to document the full options structure accepted by the parser's grade.